File: lab11/w2.py
class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        return None

    def is_empty(self):
        return len(self.items) == 0

    def peek(self):
        if not self.is_empty():
            return self.items[-1]
        return None

# Example usage: Undo functionality in a text editor
if __name__ == "__main__":
    editor_stack = Stack()
    editor_stack.push("Type 'Hello'")
    editor_stack.push("Type 'World'")
    editor_stack.push("Delete 'World'")

    print("Current state:", editor_stack.peek())  # Output: Delete 'World'
    editor_stack.pop()
    print("Undo last action. Current state:", editor_stack.peek())  # Output: Type 'World'
    editor_stack.pop()
    print("Undo last action. Current state:", editor_stack.peek())  # Output: Type 'Hello'

File: lab11/w2.cpp
#include <iostream>
#include <stack>
#include <string>

class Stack {
private:
    std::stack<std::string> items;

public:
    void push(const std::string& item) {
        items.push(item);
    }

    void pop() {
        if (!items.empty()) {
            items.pop();
        }
    }

    std::string peek() const {
        if (!items.empty()) {
            return items.top();
        }
        return "";
    }

    bool is_empty() const {
        return items.empty();
    }
};

// Example usage: Undo functionality in a text editor
int main() {
    Stack editor_stack;
    editor_stack.push("Type 'Hello'");
    editor_stack.push("Type 'World'");
    editor_stack.push("Delete 'World'");

    std::cout << "Current state: " << editor_stack.peek() << std::endl;  // Output: Delete 'World'
    editor_stack.pop();
    std::cout << "Undo last action. Current state: " << editor_stack.peek() << std::endl;  // Output: Type 'World'
    editor_stack.pop();
    std::cout << "Undo last action. Current state: " << editor_stack.peek() << std::endl;  // Output: Type 'Hello'

    return 0;
}
